library/firewall_lib.py

```diff
@@ -579,24 +746,177 @@ def exception_handler(exception_message):
             changed = True
 
     # service
-    for item in service:
-        if state == "enabled":
-            if runtime and not fw.queryService(zone, item):
-                if not module.check_mode:
-                    fw.addService(zone, item, timeout)
-                changed = True
-            if permanent and not fw_settings.queryService(item):
+    if service_operation:
```

Comment:
> This service creation and removal code should be refactored into a separate function

Reply:
My main difficulty with this is supporting check mode. I am thinking I can create a mock service object for this. Is there a better way of going about this?

Reply:
> My main difficulty with this is supporting check mode.

What specifically is the problem with check mode?

> I am thinking I can create a mock service object for this. Is there a better way of going about this?

Reply:
If you are attempting to create a service in check mode, if I understand this correctly, the managed node should not be modified. When you create a service, it directly modifies firewalld's permanent configuration, which is an issue since check mode is not supposed to modify the managed node

My way around this before was to essentially skip the code block once it determines the service specified did not previously exist and set the changed to True, which was not ideal and ended up requiring that I implemented a bunch of redundant code to handle implementing the other service element options in the case that check mode was not on.

Now that I am moving service creation to a separate function, I need to be able to handle service element functions like fw_service_settings.getDescription(), which will break if I do not create the service

TLDR: check mode should not modify managed node, moving service to its own function means that the function needs to modify the system, which contradicts the purpose of check mode.

Reply:
I can think of a couple of solutions besides a mock object now. one would just be a modification of what I had done before, which is setting changed to True and jumping over the code block if the Ansible is run in check mode and a service would have been created otherwise.

Reply:
> If you are attempting to create a service in check mode, if I understand this correctly, the managed node should not be modified. When you create a service, it directly modifies firewalld's permanent configuration, which is an issue since check mode is not supposed to modify the managed node

Correct.  check mode should not change anything
 
> My way around this before was to essentially skip the code block once it determines the service specified did not previously exist and set the changed to True, which was not ideal and ended up requiring that I implemented a bunch of redundant code to handle implementing the other service element options in the case that check mode was not on.
> 
> Now that I am moving service creation to a separate function, I need to be able to handle service element functions like fw_service_settings.getDescription(), which will break if I do not create the service

I'm still not understanding why creating a separate function for this is more difficult than having the code be in-line in the main function.  Can't you pass in the `fw_service_settings` into the function, or create it in the function and pass it back out to the main function?  Same with the `changed` boolean value and the `check_mode` parameter - pass in the `check_mode` parameter to the function, and have the function return `changed` as one of the return values (a python function can return multiple values i.e. in a tuple).

> 
> TLDR: check mode should not modify managed node, moving service to its own function means that the function needs to modify the system, which contradicts the purpose of check mode.